src/rdma.c

```diff
@@ -0,0 +1,985 @@
+        ctx->send_length = 0;
+        return VALKEY_ERR;
+    }
+
+    return VALKEY_OK;
+}
+
+
+static int rdmaSendCommand(RdmaContext *ctx, struct rdma_cm_id *cm_id, ValkeyRdmaCmd *cmd) {
+    struct ibv_send_wr send_wr, *bad_wr;
+    struct ibv_sge sge;
+    ValkeyRdmaCmd *_cmd;
+    int i;
+    int ret;
+
+    /* find an unused cmd buffer */
+    for (i = VALKEY_RDMA_MAX_WQE; i < 2 * VALKEY_RDMA_MAX_WQE; i++) {
+        _cmd = ctx->cmd_buf + i;
+        if (_cmd->keepalive.opcode == VALKEY_RDMA_INVALID_OPCODE) {
+            break;
+        }
+    }
+
+    assert(i < 2 * VALKEY_RDMA_MAX_WQE);
+
+    memcpy(_cmd, cmd, sizeof(ValkeyRdmaCmd));
+    sge.addr = (uint64_t)_cmd;
+    sge.length = sizeof(ValkeyRdmaCmd);
+    sge.lkey = ctx->cmd_mr->lkey;
+
+    send_wr.sg_list = &sge;
+    send_wr.num_sge = 1;
+    send_wr.wr_id = (uint64_t)_cmd;
+    send_wr.opcode = IBV_WR_SEND;
+    send_wr.send_flags = IBV_SEND_SIGNALED;
+    send_wr.next = NULL;
+    ret = ibv_post_send(cm_id->qp, &send_wr, &bad_wr);
+    if (ret) {
+        return VALKEY_ERR;
+    }
+
+    return VALKEY_OK;
+}
+
+static int connRdmaRegisterRx(RdmaContext *ctx, struct rdma_cm_id *cm_id) {
+    ValkeyRdmaCmd cmd = { 0 };
+
+    cmd.memory.opcode = htons(RegisterXferMemory);
+    cmd.memory.addr = htobe64((uint64_t)ctx->recv_buf);
+    cmd.memory.length = htonl(ctx->recv_length);
+    cmd.memory.key = htonl(ctx->recv_mr->rkey);
+
+    ctx->rx_offset = 0;
+    ctx->recv_offset = 0;
+
+    return rdmaSendCommand(ctx, cm_id, &cmd);
+}
+
+static int connRdmaHandleRecv(valkeyContext *c, RdmaContext *ctx, struct rdma_cm_id *cm_id, ValkeyRdmaCmd *cmd, uint32_t byte_len) {
+    if (byte_len != sizeof(ValkeyRdmaCmd)) {
+        valkeySetError(c, VALKEY_ERR_OTHER, "RDMA: FATAL error, recv corrupted cmd");
+        return VALKEY_ERR;
+    }
+
+    switch (ntohs(cmd->keepalive.opcode)) {
+    case RegisterXferMemory:
+        ctx->tx_addr = (char *)be64toh(cmd->memory.addr);
+        ctx->tx_length = ntohl(cmd->memory.length);
+        ctx->tx_key = ntohl(cmd->memory.key);
+        ctx->tx_offset = 0;
+        rdmaAdjustSendbuf(c, ctx, ctx->tx_length);
+        break;
+
+    case Keepalive:
+        break;
+
+    default:
+        valkeySetError(c, VALKEY_ERR_OTHER, "RDMA: FATAL error, unknown cmd");
+        return VALKEY_ERR;
+    }
+
+    return rdmaPostRecv(ctx, cm_id, cmd);
+}
+
+static int connRdmaHandleRecvImm(RdmaContext *ctx, struct rdma_cm_id *cm_id, ValkeyRdmaCmd *cmd, uint32_t byte_len) {
+    assert(byte_len + ctx->rx_offset <= ctx->recv_length);
+    ctx->rx_offset += byte_len;
+
+    return rdmaPostRecv(ctx, cm_id, cmd);
+}
+
+static int connRdmaHandleSend(ValkeyRdmaCmd *cmd) {
+    /* mark this cmd has already sent */
+    memset(cmd, 0x00, sizeof(*cmd));
+    cmd->keepalive.opcode = VALKEY_RDMA_INVALID_OPCODE;
+
+    return VALKEY_OK;
+}
+
+static int connRdmaHandleWrite(RdmaContext *ctx, uint32_t byte_len) {
+    UNUSED(ctx);
+    UNUSED(byte_len);
```

Comment:
Because `#define VALKEY_UNUSED __attribute__((unused))`, use `static int connRdmaHandleWrite(VALKEY_UNUSED RdmaContext *ctx, uint32_t VALKEY_UNUSED byte_len)` style instead.